From 9b4050544d7edd1f4b40b8145e1c1ac03df43889 Mon Sep 17 00:00:00 2001
From: "K.Takata" <kentkt@csc.jp>
Date: Wed, 20 Mar 2019 19:04:46 +0900
Subject: [PATCH] DirectWrite: Fix vertical alignment when 'linespace' is not 0

After 8.1.1009 (#4116), the vertical alignment of DirectWrite became slightly
different from GDI. This is because DirectWrite didn't take 'linespace' into
account. This fixes the alignment when 'linespace' is not 0.
---
 src/gui_dwrite.cpp | 2 +-
 src/gui_w32.c      | 3 ++-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/gui_dwrite.cpp b/src/gui_dwrite.cpp
index 17fa6e798a..4a24a0306f 100644
--- a/src/gui_dwrite.cpp
+++ b/src/gui_dwrite.cpp
@@ -1031,7 +1031,7 @@ DWriteContext::DrawText(const WCHAR *text, int len,
 
 	TextRenderer renderer(this);
 	TextRendererContext context = { color, FLOAT(cellWidth), 0.0f };
-	textLayout->Draw(&context, &renderer, FLOAT(x), FLOAT(y) - 0.5f);
+	textLayout->Draw(&context, &renderer, FLOAT(x), FLOAT(y));
     }
 
     SafeRelease(&textLayout);
diff --git a/src/gui_w32.c b/src/gui_w32.c
index ead617a79f..e8336b56f6 100644
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -6337,7 +6337,8 @@ gui_mch_draw_string(
 	{
 	    /* Add one to "cells" for italics. */
 	    DWriteContext_DrawText(s_dwc, unicodebuf, wlen,
-		    TEXT_X(col), TEXT_Y(row), FILL_X(cells + 1), FILL_Y(1),
+		    TEXT_X(col), TEXT_Y(row),
+		    FILL_X(cells + 1), FILL_Y(1) - p_linespace,
 		    gui.char_width, gui.currFgColor,
 		    foptions, pcliprect, unicodepdy);
 	}
